src/server/builder.ts

```diff
@@ -0,0 +1,405 @@
+/// <reference path="..\compiler\commandLineParser.ts" />
+/// <reference path="..\services\services.ts" />
+/// <reference path="protocol.d.ts" />
+/// <reference path="session.ts" />
+/// <reference types="node" />
+
+namespace ts.server {
+
+    const crypto: any = require("crypto");
+
+    /**
+     * An abstract file info that maintains a shape signature.
+     */
+    export class BuilderFileInfo {
+
+        private lastCheckedShapeSignature: string;
+        private lastEmittedContentSignature: string;
+        scriptInfo: ScriptInfo;
+
+        constructor(public fileName: string) {

```

Comment:
are there any cases when we have `BuilderFileInfo`  but matching `ScriptInfo` does not exist? If there are no any - then can we just pass `ScriptInfo` in constructor and get rid of `Builder` argument in methods for scenarios when it is not necessary? (i.e. answer of `isExternalModule` does not really depend on what project includes the file)